### PR TITLE
Implements the prefetching strategy TFPR

### LINKED ISSUE #52: Implement a prefetching strategy based on page-link analysis
### Overview

* Approach: Link analysis
* Algorithm: PageRank based
* Data to use: 
  * Page visit frequency
  * Page visit time length

### Description

Implement a prefetching strategy based on the paper [Personalized PageRank for Web Page Prediction Based on Access Time-Length and Frequency](https://dl.acm.org/doi/10.1109/WI.2007.145?download=true)

The paper describes a variation of the PageRank algorithm for conducting  Web page prediction. It uses the length of the time spent on visiting a page and how frequent a page is visited as a measure of weight.

The visiting frequency measure is already present in NAPPA.

The time spent on visiting a page can be the time between consecutive calls of the method `onResume` or the time between calling the methods `onResume` and `onStop`. See the page [Understand the Activity Lifecycle](https://developer.android.com/guide/components/activities/activity-lifecycle) for more information.

The calculation of the links follows the base PageRank formula, with the addition of the extra elements used as a weight. See the paper or the calculation summary [here](https://docs.google.com/document/d/1CQyEFeNNatXqhShBmWmNKRNk8TbaZ8PuPU1ZftYc_i8/edit#heading=h.l0txg5akttun) for more information. 

Based on what I read so far, I speculate that link analysis might make sense only for graph-like apps. For tree-like apps, the heuristics might be more appropriated.

### Adaptations required in the NAPPA context

1 Use the data from the current local user only
2 How many iterations to run?

* Test how long it takes one iteration for a given number of pages
* On average, how many iterations are needed to converge?

3 Which pages to select for the subgraph used for calculations?

* Only the children of the current node?
* All pages from the path from the root page to the children of the current page?
* Adds all ancestors of the selected pages?

4 Calculate everything in the RAM or store results in the DB?

* This answer will depend on items 2 and 3:
  * If we use all the pages of the app (known at the time) for the calculation, then it might make sense to store the scores.
  * Otherwise no, since the score depends on which pages are used in the calculation

5 How many past sessions to use?

* Using all sessions might reduce performance over time.
* Using too little sessions might yield inaccurate results. For instance, let's say only the past 2 sessions are used, the user might have different usage depending on the time of the day the app is used.
* Using too many sessions might not reflect changes in the user behaviour. Nevertheless, this does not occur frequently.

### PR DESCRIPTION
**Pull request by sshann**
Closes #52 

Inspired on the paper [Personalized PageRank for Web Page Prediction Based on Access Time-Length and Frequency](https://dl.acm.org/doi/pdf/10.1109/WI.2007.145?download=true) ([pdf](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.219.6236&rep=rep1&type=pdf))

Year: 2007
Authors:

* Yong Zhen Guo
* Kotagiri  Ramamohanarao 
* Laurence  Park 

The paper describes a link analysis approach implementation of a variation of the PageRank algorithm, where user behaviour information is considered when calculating the importance of a page:

* Access frequency:
* Access time-length

To avoid expensive computation, the calculations are performed in a subgraph instead of the entire graph.

In the NAPPA context, the algorithm described in the paper is executed at runtime, for each page navigation using data preload in memory via LiveData. The PageRank computation is executed for N iterations (**decide a default value of N**). After running N iterations, the nodes are ordered per TFPR score and the obtain all URLS that fits the URL budged, starting from the top node and descending while there is budged left.

#### Database

* Store which activity the visit time row is from. Considering the similarities with the entity `ActivityVisitTime` and `SessionData` an issue to discuss whether to merge these entities was defined as TODO message in the file [ActivityVisitTime#l11](https://github.com/S2-group/NAPPA/pull/86/files#diff-3dd52a06a063b1cd7124c4595478a80dR11) (should be created at issue #87)
* Defined a DatabaseView and select queries to obtain the aggregate visit time of the node's successors from a given node. 

These changes above are needed to correctly implement the behaviour of the algorithm described in the paper. Without the column `fromActivity`, we would be selecting the visit time from accessing the activity from any activity.

#### Prefetching Strategies

* Add new strategy and identifier key
* Add overloaded methods to serve as getters for the configuration map
* Add a method to return a flag to enable/disable fetching the visit time data 
  * Could be considered out of scope.
  * This requires refactoring for all prefetching strategies not extending from `AbstractPrefetchingStrategy` 
* Add validation to user defined parameters

#### Main Prefetch Library

* Add a LiveData list for the new select query
* Keep track of the previous activity from here
* Uses the prefetching strategy method that returns a flag to enable/disable fetching the visit time data instead of verifying the strategy key.
  * Could be considered out of scope.

#### New classes

While these could be considered out of scope, I think these additional changes simplified the workflow and avoided code duplication for this PR

* NappaConfigMap: 
  * Centralize the user-defined configuration map.
  * Ensures a single instance
  * Previously, it was located in 2 classes and we would need for another class in this PR.
* NappaThreadPool: 
  * Centralize the pool thread scheduler.
  * Ensures a single instance.
  * Previously defined in multiple locations.
  * These locations were not refactored in this PR.
* FetchSuccessorsVisitTimeHandler
  * Handles fetching the database to search the time spend in the successor's nodes when accessing the successor node from this node
  * The previous 2 classes were created already in this PR to support this handler class.
  * This Handler is part of a "proof of concept" to verify if this workflow helps to 
    * simplify the NAPPA workflow, 
    * centralize invocation to DAO classes (i.e., invocations to worker threads)
    *  reduce the responsibilities from the main PrefetchingLib class. 
  * ~~I need to do a final review, but so far I think it does.~~ After doing a code review I keep the initial judgement.

#### Gradle

* Changed minimum Android SDK level from 19 to 24 to support the following API:
  * Streams - Quick and simplified operations with arrays - SDK level 24 (since 2016)
  * Looper - API to verify if running on the main thread - SDK level 23 (since 2015)

### CODE REVIEW

File: Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/prefetch/GreedyWithPageRankScoresPrefetchingStrategy.java [hunk 1]
```diff
@@ -29,6 +29,10 @@ public GreedyWithPageRankScoresPrefetchingStrategy(float threshold) {
         this.threshold = threshold;
     }
 
+    @Override
+    public boolean needVisitTime() {
+        return false;
+    }
```
Comment (sshann) [inline Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/prefetch/GreedyWithPageRankScoresPrefetchingStrategy.java:35]: These are only needed when not extending the class `AbstractPrefetchingStrategy`. However, to extend this class we need to do a small refactoring, which is outside the context of this PR

File: Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/handler/activity/visittime/FetchSuccessorsVisitTimeHandler.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * {@link nl.vu.cs.s2group.nappa.room.activity.visittime.ActivityVisitTime ActivityVisitTime}
+ * for each successor of the provided node. After fetching the data, this handler will
+ * invoke the method {@link ActivityNode#setSuccessorsAggregateVisitTimeLiveData(LiveData)}
+ * for the provided node.
+ */
+public class FetchSuccessorsVisitTimeHandler {
+    private static final String LOG_TAG = FetchSuccessorsVisitTimeHandler.class.getSimpleName();
+
+    public static void run(ActivityNode activity) {
+        if (Looper.getMainLooper().isCurrentThread())
```
Comment (sshann) [inline Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/handler/activity/visittime/FetchSuccessorsVisitTimeHandler.java:31]: Requires SDK level 23

File: Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/util/NappaUtil.java [hunk 3]
```diff
@@ -120,4 +120,33 @@ public static int getSuccessorsTotalAggregateVisitFrequency(@NotNull ActivityNod
...
+     * succeeding activity from the {@code sourceNode} activity.
+     *
+     * @param sourceNode The activity to use as source.
+     * @return The total aggregate time.
+     */
+    public static long getSuccessorsAggregateVisitTimeOriginatedFromNode(@NonNull ActivityNode sourceNode) {
+        return sourceNode.getSuccessorsVisitTimeList()
+                .stream()
+                .mapToLong(visitTime -> visitTime.totalDuration)
+                .sum();
```
Comment (sshann) [inline Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/util/NappaUtil.java:135]: SDK level 24

File: Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/prefetch/TfprPrefetchingStrategy.java [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+          * The long time was increased to a few minutes;
+          * The current node had a single child only;
+
+          In both cases the TFPR score was between 60% and 75%.
+
+          Since this is not always the case, it seems reasonable to reduce the lower threshold.
+          The value is only overridden if it was not already overridden using the user-defined
+          configurations.
+         */
```
Comment (sshann) [inline Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/prefetch/TfprPrefetchingStrategy.java:62]: There is the question if this behaviour will occur during the experiments and/or in real usage scenario,

Comment (iivanoo) [reply Prefetching-Library/android_prefetching_lib/src/main/java/nl/vu/cs/s2group/nappa/prefetch/TfprPrefetchingStrategy.java:62]: You can try with a single run of the experiment over all subjects and see how the scores fluctuate, ok?

**Pull request merged**